File: 977.squares-of-a-sorted-array.java
/*
 * @lc app=leetcode id=977 lang=java
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
class Solution {
    public int[] sortedSquares(int[] nums) {
        if (nums.length == 1) {
            return new int[] { nums[0] * nums[0] };
        }
        int[] ret = new int[nums.length];
        int idx = nums.length - 1;
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            if (Math.abs(nums[left]) > Math.abs(nums[right])) {
                ret[idx--] = nums[left] * nums[left];
                left += 1;
            } else {
                ret[idx--] = nums[right] * nums[right];
                right -= 1;
            }
        }
        return ret;
    }
}
// @lc code=end

File: 977.squares-of-a-sorted-array.cpp
/*
 * @lc app=leetcode id=977 lang=cpp
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start

#include <vector>

using namespace std;

class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> ret;
        auto inserter = back_inserter(ret);
        auto lptr = nums.begin(), rptr = nums.end();
        while (lptr < rptr) {
            auto lval = abs(*lptr), rval = abs(*(rptr - 1));
            if (lval > rval) {
                *inserter = lval * lval;
                lptr += 1;
            } else {
                *inserter = rval * rval;
                rptr -= 1;
            }
        }
        reverse(ret.begin(), ret.end());
        return ret;
    }
};
// @lc code=end
